pennylane/templates/state_preparations/mottonen.py

```diff
@@ -87,36 +87,48 @@ def _compute_theta(alpha):
     k = np.log2(alpha.shape[0])
     factor = 2 ** (-k)
 
-    theta = sparse.dok_matrix(alpha.shape, dtype=np.float64)  # type: sparse.dok_matrix
+    theta = np.zeros(alpha.shape, dtype=np.float64)
 
     for row in range(alpha.shape[0]):
         # Use transpose of M:
-        entry = sum([_matrix_M_entry(col, row) * a for (col, _), a in alpha.items()])
+        entry = sum(_matrix_M_entry(col, row) * alpha[col] for col in range(len(alpha)))
         entry *= factor
         if abs(entry) > 1e-6:
-            theta[row, 0] = entry
+            theta[row] = entry
 
     return theta
 
 
 def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
-    """Applies a given inverse rotation to the target qubit
-    that is uniformly controlled by the control qubits.
+    """Applies a uniformly-controlled rotation to the target qubit.
+
+    A uniformly-controlled rotation is a sequence of multi-controlled
+    rotations, each of which is conditioned on the control qubits being in a different state.
+    For example, a uniformly-controlled rotation with two control qubits decomposes into
+    four multi-controlled rotations, each applying the rotation only if the control qubits
+    are in states :math:`|00\rangle`, :math:`|01\rangle`, :math:`|10\rangle`, and :math:`|11\rangle`, respectively.
+    For more details, see Möttönen et al. (2004) <https://arxiv.org/pdf/quant-ph/0407010.pdf>`_
+
+    To implement a uniformly-controlled rotation, a decomposition based on gray codes is
+    used. For this purpose, the multi-controlled rotation angles alpha have to be converted into
+    a different set of rotation angles.
+
+    For more details, see `Möttönen and Vartiainen (2005), Fig 7a<https://arxiv.org/pdf/quant-ph/0504100.pdf>`_.
 
     Args:
-        gate (~.Operation): gate to be applied, needs to have exactly
-            one parameter
-        alpha (array[float]): alpha parameters
+        gate (~.Operation): gate to be applied, needs to have exactly one parameter
+        alpha (array[float]): angles to decompose the uniformly-controlled rotation into multi-controlled rotations
         control_wires (array[int]): wires that act as control
         target_wire (int): wire that acts as target
     """
 
-    theta = _compute_theta(alpha)  # type: sparse.dok_matrix
+    theta = _compute_theta(alpha)
 
     gray_code_rank = len(control_wires)
 
     if gray_code_rank == 0:
-        gate(theta[0, 0], wires=[target_wire])
+        if theta[0] != 0.0:
```

Comment:
I added this and a similar if statement below, because the case that `theta[0] = 0` happens quite often and does not require a rotation!

Reply:
So these are guaranteed to be exactly 0 as per lines 97 and 98 above right? It would be interesting to analyze the tradeoff between the cutoff point, and accuracy / execution time.

Reply:
Good point, I just took this from above and did not think of precision. We haven't heard any issues with this yet but might need to change this behaviour for differentiability...